### PR TITLE
build: don't produce thin archives

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

**Comment by ry** [comment]
Can you also build a new `v8.lib` for prebuilt, so we can verify that this change is actually fixing a bug?

File: tools/setup.py [hunk 1]
```diff
@@ -137,6 +132,17 @@ def generate_gn_args(mode):
...
+    # On Windows, we need to use a custom toolchain, for the following reasons:
+    #   * When building a static library, the default chromium toolchain
+    #     produces 'thin archives' when building a static library. These are not
+    #     compatible with Microsoft's link.exe, and therefore rustc can't use
+    #     such libraries out of the box.
+    #   * The default toolchain doesn't support cc_wrapper.
+    # TODO: Is there a way to use it without the involvement of args.gn?
+    if os.name == "nt":
+        tc = "//build_extra/toolchain/win:win_clang_x64"
+        out += ['custom_toolchain="%s"' % tc, 'host_toolchain="%s"' % tc]
```
Comment (ry) [inline tools/setup.py]: Configuration shouldn't be in setup.py - configuration should be in gn files - that's the purpose. I should have said this before when the above configurations were added here.
 We should fork `BUILDCONFIG.gn` and set the toolchain other settings in this file there. Alternatively you can start `tools/args_win.gn` and copy that into place.

File: build_extra/toolchain/win/BUILD.gn [hunk 2]
```diff
@@ -111,7 +111,7 @@ template("msvc_toolchain") {
 
       # lld-link includes a replacement for lib.exe that can produce thin
       # archives and understands bitcode (for lto builds).
-      lib = "$prefix/$lld_link /lib /llvmlibthin"
+      lib = "$prefix/$lld_link /lib"
```
Comment (ry) [inline build_extra/toolchain/win/BUILD.gn]: Add a gn flag for this, and we should send the patch upstream. I'd very much like to not diverge from chrome's build.

File: build_extra/toolchain/validate.gni [hunk 3]
```diff
@@ -2,18 +2,18 @@
...
-    print("The 'cc_wrapper' option isn't supported by the default Windows" +
-          " toolchain. To make it work, add these gn arguments:")
-    print("  custom_toolchain=\"$cc_wrapper_toolchain\"")
-    print("  host_toolchain=\"$cc_wrapper_toolchain\"")
-    assert(toolchain_supports_cc_wrapper)
+    print("The Windows build requires a custom toolchain. " +
+          "Add these gn arguments:")
+    print("  custom_toolchain=\"$win_toolchain\"")
+    print("  host_toolchain=\"$win_toolchain\"")
+    assert(win_toolchain_in_use)
```
Comment (ry) [inline build_extra/toolchain/validate.gni]: Remove validate.gn after you've forked BUILDCONFIG.gn (or other such way of not relying on setup.py)

File: tools/setup.py (hunk 1 above)
Comment (piscisaureus) [reply tools/setup.py]: Sorry, not possible without forking //build.

**Comment by ry** [comment]
Are you able to link to cargo built rust objects now?

File: build_extra/toolchain/win/BUILD.gn (hunk 2 above)
Comment (piscisaureus) [reply build_extra/toolchain/win/BUILD.gn]: > Add a gn flag for this, and we should send the patch upstream. 
> I'd very much like to not diverge from chrome's build.

Those comments are in direct conflict with one another. This BUILD.gn is derived from chromiums and modified with some simple sed edits. You can opt to either full-on fork it or you can live without the flag for now.

**Comment by piscisaureus** [comment]
> Are you able to link to cargo built rust objects now?

Yes, works smoothly.

**Comment by piscisaureus** [comment]
Closing since review comments can't be addressed.

File: tools/setup.py (hunk 1 above)
Comment (ry) [reply tools/setup.py]: can you explain? can't you just change this line?
https://github.com/denoland/deno/blob/master/.gn#L6

**Pull request closed without merging**